rs/registry/node_provider_rewards/src/v1_rewards.rs

```diff
@@ -0,0 +1,535 @@
+
+use crate::v1_logs::LogLevel;
+use crate::v1_types::TimestampNanos;
+use crate::{
+    v1_logs::{LogEntry, Operation, RewardsLog},
+    v1_types::{
+        DailyNodeMetrics, RegionNodeTypeCategory, RewardableNode, Rewards, RewardsPerNodeProvider,
+    },
+};
+use ic_management_canister_types::{NodeMetrics, NodeMetricsHistoryResponse};
+use rust_decimal::Decimal;
+use rust_decimal_macros::dec;
+use std::collections::HashMap;
+
+const FULL_REWARDS_MACHINES_LIMIT: u32 = 3;
+const MIN_FAILURE_RATE: Decimal = dec!(0.1);
+const MAX_FAILURE_RATE: Decimal = dec!(0.6);
+const MAX_REWARDS_REDUCTION: Decimal = dec!(0.8);
+const RF: &str = "Linear Reduction factor";
+
+pub fn calculate_rewards(
+    days_in_period: u64,
+    rewards_table: &NodeRewardsTable,
+    subnet_metrics: HashMap<PrincipalId, Vec<NodeMetricsHistoryResponse>>,
+    rewardable_nodes: &[RewardableNode],
+) -> RewardsPerNodeProvider {
+    let mut rewards_per_node_provider = HashMap::default();
+    let mut rewards_log_per_node_provider = HashMap::default();
+
+    let mut all_assigned_metrics = daily_node_metrics(subnet_metrics);
+    let subnets_systematic_fr = systematic_fr_per_subnet(&all_assigned_metrics);
+    let node_provider_rewardables = rewardables_by_node_provider(rewardable_nodes);
+
+    for (node_provider_id, node_provider_rewardables) in node_provider_rewardables {
+        let mut logger = RewardsLog::default();
+        logger.add_entry(
+            LogLevel::High,
+            LogEntry::CalculateRewardsForNodeProvider(node_provider_id),
+        );
+
+        let assigned_metrics: HashMap<PrincipalId, Vec<DailyNodeMetrics>> =
+            node_provider_rewardables
+                .iter()
+                .filter_map(|node| {
+                    all_assigned_metrics
+                        .remove(&node.node_id)
+                        .map(|daily_metrics| (node.node_id, daily_metrics))
+                })
+                .collect::<HashMap<PrincipalId, Vec<DailyNodeMetrics>>>();
+        let node_daily_fr =
+            nodes_idiosyncratic_fr(&mut logger, &assigned_metrics, &subnets_systematic_fr);
+
+        let rewards = node_provider_rewards(
+            &mut logger,
+            &node_provider_rewardables,
+            node_daily_fr,
+            days_in_period,
+            rewards_table,
+        );
+
+        rewards_log_per_node_provider.insert(node_provider_id, logger);
+        rewards_per_node_provider.insert(node_provider_id, rewards);
+    }
+
+    RewardsPerNodeProvider {
+        rewards_per_node_provider,
+        rewards_log_per_node_provider,
+    }
+}
+
+/// Computes the idiosyncratic daily failure rates for each node.
+///
+/// This function calculates the idiosyncratic failure rates by subtracting the systematic
+/// failure rate of the subnet from the node's failure rate for each day.
+/// If the node's failure rate is less than the systematic failure rate, the idiosyncratic
+/// failure rate is set to zero.
+fn nodes_idiosyncratic_fr(
```

Comment:
idiosyncratic has more of a connotation of "weird" or "peculiar", and maybe only serves to tell someone to read the function.

Maybe we could say "node_failure_rate_relative_to_subnet" or something more wordy, especially since it's only used locally.  I'm not sure that's a great name either.